Batteries/Control/OptionT.lean

```diff
@@ -0,0 +1,95 @@
+/-
+Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+-/
+
+import Batteries.Control.Lemmas
+
+/-!
+# Lemmas About Option Monad Transformer
+
+This file contains lemmas about the behavior of `OptionT` and `OptionT.run`.
+-/
+
+universe u v
+
+namespace OptionT
+
+variable {α β : Type u} {m : Type u → Type v} (x : OptionT m α)
+
+@[ext] theorem ext {x x' : OptionT m α} (h : x.run = x'.run) : x = x' := h
+
+@[simp] theorem run_mk (x : m (Option α)) : OptionT.run (OptionT.mk x) = x := rfl
+
+section Monad
+
+variable [Monad m]
+
+@[simp] theorem run_pure (a) : (pure a : OptionT m α).run = pure (some a) := rfl
+
+@[simp]
+theorem run_bind (f : α → OptionT m β) :
+    (x >>= f).run = x.run >>= fun
+                              | some a => OptionT.run (f a)
+                              | none   => pure none := rfl
+
+@[simp]
+theorem run_map (f : α → β) [LawfulMonad m] : (f <$> x).run = Option.map f <$> x.run := by
+  rw [← bind_pure_comp _ x.run]
+  change x.run >>= (fun
+                     | some a => OptionT.run (pure (f a))
+                     | none   => pure none) = _
+  apply bind_congr
+  intro a; cases a <;> simp [Option.map, Option.bind]
+
+@[simp] theorem run_monadLift {n} [MonadLiftT n m] (x : n α) :
+    (monadLift x : OptionT m α).run = (monadLift x : m α) >>= fun a => pure (some a) := rfl
```

Comment:
Or maybe
```suggestion
    (monadLift x : OptionT m α).run = some <$> (monadLift x : m α) := rfl
```
(untested)

Reply:
This does require the base monad `m` be lawful but I don't think that's a very important generalization, and the map seems cleaner so I've changed it to that